src/toil/provisioners/aws/awsProvisioner.py

```diff
@@ -0,0 +1,325 @@
+import subprocess
+import time
+import logging
+from boto.ec2.blockdevicemapping import BlockDeviceMapping, BlockDeviceType
+from boto.exception import BotoServerError
+from cgcloud.lib.ec2 import ec2_instance_types, retry_ec2
+from itertools import islice
+
+from toil.batchSystems.abstractBatchSystem import AbstractScalableBatchSystem
+from toil.provisioners.abstractProvisioner import AbstractProvisioner, Shape
+from toil.provisioners.cgcloud.provisioner import CGCloudProvisioner
+from toil.provisioners.aws import *
+from cgcloud.lib.context import Context
+from boto.utils import get_instance_metadata
+
+
+logger = logging.getLogger(__name__)
+
+
+def expectedLaunchErrors(e):
+    return e.status == 400 and ('iamInstanceProfile.arn is invalid' in e.body or 'has no associated IAM' in e.body)
+
+
+def expectedShutdownErrors(e):
+    return e.status == 400 and 'dependent object' in e.body
+
+
+class AWSProvisioner(AbstractProvisioner):
+
+    def __init__(self, config, batchSystem):
+        self.ctx = Context(availability_zone='us-west-2a', namespace='/')
+        self.instanceMetaData = get_instance_metadata()
+        self.securityGroupName = self.instanceMetaData['security-groups']
+        self.spotBid = None
+        parsedBid = config.nodeType.split(':', 1)
+        if len(config.nodeType) != len(parsedBid[0]):
+            # there is a bid
+            self.spotBid = parsedBid[1]
+            self.instanceType = ec2_instance_types[parsedBid[0]]
+        else:
+            self.instanceType = ec2_instance_types[config.nodeType]
+        self.batchSystem = batchSystem
+        self.masterIP = self.instanceMetaData['local-ipv4']
+        self.keyName = self.instanceMetaData['public-keys'].keys()[0]
+
+    def setNodeCount(self, numNodes, preemptable=False, force=False):
+        # get all nodes in cluster
+        workerInstances = self._getWorkersInCluster(preemptable)
+        instancesToLaunch = numNodes - len(workerInstances)
+        logger.info('Adjusting cluster size by %s', instancesToLaunch)
+        if instancesToLaunch > 0:
+            self._addNodes(instancesToLaunch, preemptable=preemptable)
+        elif instancesToLaunch < 0:
+            self._removeNodes(instances=workerInstances, numNodes=numNodes, preemptable=preemptable, force=force)
+        else:
+            pass
+        workerInstances = self._getWorkersInCluster(preemptable)
+        return len(workerInstances)
+
+    def getNodeShape(self, preemptable=False):
+        instanceType = self.instanceType
+        return Shape(wallTime=60 * 60,
+                     memory=instanceType.memory * 2 ** 30,
+                     cores=instanceType.cores,
+                     disk=(instanceType.disks * instanceType.disk_capacity * 2 ** 30))
+
+    @staticmethod
+    def sshCluster(securityGroupName):
+        ctx = Context(availability_zone='us-west-2a', namespace='/')
+        instances = AWSProvisioner.__getNodesInCluster(ctx, securityGroupName, both=True)
```

Comment:
I've done the conversion from `@staticmethod` to `@classmethod` so often that I simply default to @classmethod now. I turned off the "this method could be static" warning in PyCharm.